### DISCUSSION TITLE
How do I know if the terminal is running a command?

### DISCUSSION THREAD

**Discussion by CaledoniaProject** [sourcelair/xterm.js#4575] [Q&A] [answered]
In the webUI I have a button to input commands to the xterm automatically, I want to know if the console is "busy", so that I won't inject the input sequence and tell user what's happening.

How can I know if the console is busy? e.g running vim

**Comment by jerch** [comment, marked as answer]
> How can I know if the console is busy? e.g running vim

Well - you cant. A terminal is just an IO device for interactive keyboard input and screen output, where the OS redirects process `stdin` and `stdout` to (through TTY/PTY abstraction).

_some explanation_
The commands are launched by another process, typically a shell. It is important to understand that terminal != shell - if you want control/knowledge about the processes being spawned and attached to the TTY, you'd have to hook into the shell's command evaluation (the E in REPL). With an interactive shell + command execution the situation gets further complicated by the fact, that commands can be interactive itself (its own REPL) temporarily replacing the shell as foreground REPL, with its very own idea of commands (thats what makes it possible to run python commands at the python REPL w'o interfering with your shell). This may lead to a tree of (repl) processes, where all or only the last one has read/write access to the TTY.
In the same sense vim is just a REPL with its own command set to manipulate file contents and showing/giving you an interactive feedback about your changes. While a shell normally knows, which processes it started and when they finally return (not for subprocesses!), it does not know the commend set being used by that process (`vim` could be something else). The terminal on the other hand knows nothing about that at all. It just receives screen update commands from the OS (think of it like a "textual GPU", you also would not ask the GPU, who currently draws at position xy).

So nope - it is conceptually not possible, as any of the processes might read/write to the terminal. The TTY as found on POSIX systems is simply not made for that. Only the kernel knows all the needed the details in the end, so well - it might be possible with a kernel module hooking into TTY abstraction :see_no_evil:.
